src/controller/CHIPDeviceController.cpp

```diff
@@ -268,6 +268,13 @@ void DeviceController::OnSessionReleased(SessionHandle session)
     mCASESessionManager->OnSessionReleased(session);
 }
 
+void DeviceController::OnFirstMessageDeliveryFailed(SessionHandle session)
```

Comment:
I think it's wrong that in many places `SessionHandle` is passed by value. It's quite a big object. Could you use `const SessionHandle&`?